[APM] Adds new metrics charts for java agent: GC time and GC rate

ogupte opened the pull request:
<img width="1345" alt="Screen Shot 2019-06-17 at 4 19 47 PM" src="https://user-images.githubusercontent.com/1967266/59642925-bfd40b00-911b-11e9-9ccb-5019d78ee715.png">

ogupte committed:
Adds new metrics charts for java agent: GC time and GC rate

ogupte commented:
Bar series chart seems to make the most sense here as a visualization of total events / time, but i also tried rendering as linemark and area series charts:
<img width="1359" alt="Screen Shot 2019-06-11 at 10 42 23 PM" src="https://user-images.githubusercontent.com/1967266/59329688-53cd4f00-8ca4-11e9-8658-df42ab552959.png">
<img width="1351" alt="Screen Shot 2019-06-11 at 11 17 55 PM" src="https://user-images.githubusercontent.com/1967266/59329689-5465e580-8ca4-11e9-8e74-75470fd4c7e5.png">
How do folks feel about the GC metrics are visualized?

eyalkoren commented:
@ogupte Looks good!
Few initial comments:
1. The GC graphs show deltas, so it would make sense to exclude every measurement that doesn't have preceding measurement, instead of assigning 0 to the preceding. I assume this is the reason for the initial spikes?
2. We need to work on GC labeling- make them clearer, eg- "GC cycles per minute" and "GC ms per minute" (obviously those are not very good ones, but something in this direction)

roncohen commented:
Thanks @ogupte! 

The derivative for `jvm.gc.count` and `jvm.gc.time` here isn’t going to be correct because we send a metrics series for each generation. So we’re potentially making a derivative between to unrelated metrics series. This makes it difficult to to a total gc time graph. We can show the individual generations on a graph. Alternatively, I think we need to think a bit more on the metrics the java agent sends up.

And one for when you get to polishing mode: Counts don't need a decimal point (collection rate), unless it's a per second rate?

felixbarny commented:
Rather than having a separate graph for the gc rate (count), couldn't we add the counts per bucket/second to the tooltip?

eyalkoren commented:
Adding to what Ron commented:
The different GCs are identified based on the tags sent with them.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1584/)

eyalkoren commented:
@felixbarny and I had a long discussion regarding this now. Here is some output:

- We agreed that having a graph per GC doesn't make sense, especially because we have no guarantee regarding the number of GCs employed by a JVM, or worse- multiple JVMs potentially of different versions (which may use different names for the GCs). So multiple lines on one graph should be a better option.
- Regarding metrics from multiple JVMs presented on the same graph: we need to decide what we do with that- one option is two lines per GC- average and max. This may not be very easy to read and may result in too many lines on a single graph. Another option is only a line of max per GC, probably better than average (?)
- We need better captioning, as discussed above
- Do we want to use derivatives here? They suffer from the same drawbacks as discussed with the breakdown metrics. Do we want to use derivatives sometimes and deltas in other times? Reporting deltas by the agent should not be very difficult, it practically only means keeping the last value reported. We had discussions about that in the past, can't quite recall what were the objections around that..

ogupte committed:
[APM] update Java GC metrics based on feedback: added another level of
aggregation over labels.name and renders each group in its own line
series on the charts. (Old and yound generation GCs) also updated chart
style and labels

ogupte commented:
I added another aggregation over `labels.name` so now the old and young generation are both plotted on the same chart with different lines. I also got rid of `derivative` in favor of `serial_diff`. I updated the chart labels to include value of `labels.name` and changed the formatting of the values to integers. Let me know what you think.

felixbarny commented:
Cool!

Which aggregation (sum, max avg) is used for the GC graphs? I feel like that should be part of the legend.
Can the tooltips be sorted by average value rather than alphabetically? The Old Gen is the first label while it's the lowest graph.
Are the values per bucket or are they normalized to a specific time frame, like collection time/minute? From experience, the sum aggregation does not play well with a per-bucket scaling because as you increase the time range, the values increase.

formgeist commented:
I'm getting TypeErrors trying to pull up the Metrics page for Java

```TypeError: Cannot read property 'gcCountAll' of undefined
    at Object.keys.map (/Users/casperhubertz/dev/elastic/kibana/x-pack/plugins/apm/server/lib/metrics/transform_metrics_chart.ts:141:29)
    at Array.map (<anonymous>)
    at map (/Users/casperhubertz/dev/elastic/kibana/x-pack/plugins/apm/server/lib/metrics/transform_metrics_chart.ts:134:44)
    at Array.reduce (<anonymous>)
    at reduce (/Users/casperhubertz/dev/elastic/kibana/x-pack/plugins/apm/server/lib/metrics/transform_metrics_chart.ts:121:34)
    at getGCRateChartData (/Users/casperhubertz/dev/elastic/kibana/x-pack/plugins/apm/server/lib/metrics/by_agent/java/gc_rate/index.ts:32:10)
    at process._tickCallback (internal/process/next_tick.js:68:7)```

I can run this off of the `master` snapshot , right?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2048/)

ogupte committed:
[APM] create 2 aggregations (max,avg) per generation label for each of
the Java GC metrics charts (GC activity & GC time). Clean up of agg
queries.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2472/)

roncohen commented:
hey @ogupte. Great work on this so far. I really like the bar charts you showed above!

We've been discussing how to get this over the finish line, the issues around displaying data from  multiple JVMs in the same graph, multiple generations etc. For now, let's do this:

- [ ] 1. For number of GC collections and time spent in GC we will show a graph per generation. It should be dynamic so that if there are more generations, it will show more graphs. Each graph should show two lines and avg. and a max. As an example, if there are two generations, we show 2x GC collections graphs and 2x GC time spent graphs. Each of these 4 graphs should have two lines, avg. and max.
- [ ] 2. The graphs should be normalized to per minute. That is, even if the bucket size changes, the numbers should reflect the per minute number.
- [ ] 3. Legends and popovers should be clearly labeled with their aggregate function and time unit, e.g. `avg. GC time ms / min` or `avg. GC collections / min`

ogupte commented:
Closing this for now. Will reopen in a new PR that implements a newer design.
